koog-ktor/src/jvmMain/kotlin/ai/koog/ktor/KoogKtorServerPluginJvm.kt

```diff
@@ -0,0 +1,131 @@
+package ai.koog.ktor
+
+import ai.koog.agents.mcp.DefaultMcpToolDescriptorParser
+import ai.koog.agents.mcp.McpToolDescriptorParser
+import ai.koog.agents.mcp.McpToolRegistryProvider
+import ai.koog.agents.mcp.McpToolRegistryProvider.DEFAULT_MCP_CLIENT_NAME
+import ai.koog.agents.mcp.McpToolRegistryProvider.DEFAULT_MCP_CLIENT_VERSION
+import io.modelcontextprotocol.kotlin.sdk.client.Client
+import io.modelcontextprotocol.kotlin.sdk.shared.DEFAULT_REQUEST_TIMEOUT
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withTimeout
+import kotlin.time.Duration
+
+/**
+ * Configuration class for MCPTools that manages the integration of various tool registries
+ * into the system. Provides methods to process tools using different transport mechanisms.
+ *
+ * @param agentConfig Configuration for the Koog agent server, which includes tool registry details.
+ */
+public class MCPToolsConfig(
+    private val agentConfig: KoogAgentsConfig.AgentConfig,
+    private val scope: CoroutineScope
+) {
+    private val mutex = Mutex()
```

Comment:
Why do you need a `Mutex` here?
Looks like it brings a lot of coroutines complexity for no benefit at all (`runBlocking`, parent-less `Job`, scopes, new coroutines for updating `agentConfig`). To me, this looks like the `agentConfig` is only used synchronously during Application startup, which eliminates races. But even if it is not, and someone configures this in parallel - then, first of all, it's their problem, don't do this, and second - we can just use a `ConcurrentHashMap` under the hood for such cases. (and `synchronous` methods for non-JVM targets)

Reply:
> Looks like it brings a lot of coroutines complexity for no benefit at all (runBlocking, parent-less Job, scopes, new coroutines for updating agentConfig).

The problem is that we need to use a `suspend` API from Koog/MCP SDK.. So a `CoroutineScope` is needed as a minimum. Or, we need to drop support for this in the plugin altogether. Which I guess we could do, and/or postpone it to a later PR.

Looking at the code again, I think we can make this _async_ by just launching everything, but awaiting it before the first agent actually kicks-off. The only _proper_ scope we can use is `Application`, but I don't have access to that from a Routing based plugin ...

WDYT?

Reply:
What suspend API?

Reply:
I ended up with a mix.. When the plugin is installed I get the `Application` and use that create a proper-scope, and use that to lazily initialise the MCP connections in parallel. When the first agent is created the MCP connections are awaited properly without blocking.